contracts/asset-proxy/contracts/src/ERC1155Proxy.sol

```diff
@@ -0,0 +1,268 @@
+                let assetDataOffset := calldataload(4)
+
+                // Load length in bytes of `assetData`, computed by:
+                // 4 (function selector)
+                // + assetDataOffset
+                let assetDataLength := calldataload(add(4, assetDataOffset))
+
+                // This corresponds to the beginning of the Data Area for Table #3.
+                // Computed by:
+                // 4 (function selector)
+                // + assetDataOffset
+                // + 32 (length of assetData)
+                calldatacopy(32, add(36, assetDataOffset), assetDataLength)
+
+                // Increment by 32 the offsets to `tokenIds`, `tokenValues`, and `callbackData` 
+                mstore(68, add(mload(68), 32))
+                mstore(100, add(mload(100), 32))
+                mstore(132, add(mload(132), 32))
+
+                // Record the address of the destination erc1155 asset for later.
+                let assetAddress := and(
+                    mload(36),
+                    0x000000000000000000000000ffffffffffffffffffffffffffffffffffffffff
+                )
+
+                ////////// STEP 2/4 //////////
+                let scaleAmount := calldataload(100)
+                let tokenValuesOffset := add(mload(100), 4) // add 4 for calldata offset
+                let tokenValuesLengthInBytes := mul(
+                    mload(tokenValuesOffset),
+                    32
+                )
+                let tokenValuesBegin := add(tokenValuesOffset, 32)
+                let tokenValuesEnd := add(tokenValuesBegin, tokenValuesLengthInBytes)
+                for { let tokenValueOffset := tokenValuesBegin }
+                    lt(tokenValueOffset, tokenValuesEnd)
+                    { tokenValueOffset := add(tokenValueOffset, 32) }
+                {
+                    // Load token value and generate scaled value
+                    let tokenValue := mload(tokenValueOffset)
+                    let scaledTokenValue := mul(tokenValue, scaleAmount)
+                    
+                    // Check if scaled value is zero
+                    if iszero(scaledTokenValue) {
+                        // Revert with `Error("TRANSFER_GREATER_THAN_ZERO_REQUIRED")`
+                        mstore(0, 0x08c379a000000000000000000000000000000000000000000000000000000000)
+                        mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
+                        mstore(64, 0x000000235452414e534645525f475245415445525f5448414e5f5a45524f5f52)
+                        mstore(96, 0x4551554952454400000000000000000000000000000000000000000000000000)
+                        mstore(128, 0)
+                        revert(0, 132)
+                    }
+
+                    // Check for multiplication overflow
+                    let expectedTokenValue := div(scaledTokenValue, scaleAmount)
+                    if iszero(eq(expectedTokenValue, tokenValue)) {
```

Comment:
Just a note that if we do remove the last check, we'll still need to check if `scaleAmount` is 0 before doing this division.

Reply:
A check has been added above, right after we load `amount`.